src/Npgsql/Internal/Resolvers/AdoTypeInfoResolver.cs

```diff
@@ -470,21 +471,44 @@ public AdoArrayTypeInfoResolver()
     public new PgTypeInfo? GetTypeInfo(Type? type, DataTypeName? dataTypeName, PgSerializerOptions options)
     {
         var info = Mappings.Find(type, dataTypeName, options);
-        if (info is null && dataTypeName is not null)
-            info = GetEnumArrayTypeInfo(type, dataTypeName.GetValueOrDefault(), options);
+
+        Type? elementType = null;
+        if (info is null && dataTypeName is not null &&
+            (type is null || type == typeof(object) || TypeInfoMappingCollection.IsArrayLikeType(type, out elementType))
+            && options.DatabaseInfo.GetPostgresType(dataTypeName) is PostgresArrayType { Element: var pgElementType })
+        {
+            info = GetEnumArrayTypeInfo(elementType, pgElementType, type, dataTypeName.GetValueOrDefault(), options) ??
+                   GetObjectArrayTypeInfo(elementType, pgElementType, type, dataTypeName.GetValueOrDefault(), options);
+        }
         return info;
     }
 
-    static PgTypeInfo? GetEnumArrayTypeInfo(Type? type, DataTypeName dataTypeName, PgSerializerOptions options)
+    static PgTypeInfo? GetObjectArrayTypeInfo(Type? elementType, PostgresType pgElementType, Type? type, DataTypeName dataTypeName,
+        PgSerializerOptions options)
     {
-        if (type is not null && type != typeof(object) && (!TypeInfoMappingCollection.IsArrayLikeType(type, out var elementType) || elementType != typeof(string)))
+        if (elementType != typeof(object))
+            return null;
+
+        // Probe if there is any mapping at all for this element type.
+        var elementId = options.ToCanonicalTypeId(pgElementType);
+        if (options.GetDefaultTypeInfo(elementId) is null)
             return null;
 
-        if (options.DatabaseInfo.GetPostgresType(dataTypeName) is not PostgresArrayType { Element: PostgresEnumType enumType })
+        var mappings = new TypeInfoMappingCollection();
+        mappings.AddType<object>(pgElementType.DataTypeName,
```

Comment:
For which case is this?

Reply:
For the subsequent mappings.AddArrayType to succeed it needs to be able to find the element mapping.

For reference types you could manually create the mapping too (and then pass it into the overload). More generally for dynamic resolutions (see `DynamicTypeInfoResolver`) it's not so convenient. When dealing with structs a single add produce multiple mappings (T and T?) that all need to be picked up to create the array mapping.